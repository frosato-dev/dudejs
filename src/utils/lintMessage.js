import load from "@commitlint/load/lib/index";
import lint from "@commitlint/lint/lib/index";

export default async commitMessage => {
  const CONFIG = require("./../configs/commitlint");

  const options = await load(CONFIG);
  const report = await lint(commitMessage, options.rules);

  return report.errors;
};
