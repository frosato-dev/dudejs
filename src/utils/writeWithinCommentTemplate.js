export default commands => `
#### DudeJS - START ####
${commands.join("\n")}
#### DudeJS - END ####
`;
