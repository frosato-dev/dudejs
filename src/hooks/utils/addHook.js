import path from "path";
import fs from "fs";
import readPkgUp from "read-pkg-up";
import writeWithinCommentTemplate from "./../../utils/writeWithinCommentTemplate";

const modulePath = fs.realpathSync(process.cwd());
const clientProjectPath = path.join(modulePath, "../..");
const gitHookPath = path.join(clientProjectPath, ".git", "hooks");

export default (hookName, commands = []) => {
  const { pkg: packageJson } = readPkgUp.sync({ cwd: clientProjectPath });

  if (!packageJson) {
    //eslint-disable-next-line
    console.log(
      `‚ùå DudeJS was not able to add ${hookName} hook, come on dude did you forget to git init?`,
    );
    process.exit();
  }

  const filePath = path.join(gitHookPath, hookName);

  if (fs.existsSync(filePath)) {
    //eslint-disable-next-line
    console.log(
      `üòé Seems you already have a ${hookName} hook, DudeJS hasn't done anything and he likes it.`,
    );
    process.exit();
  }

  const fileContent = writeWithinCommentTemplate(commands);
  fs.writeFileSync(filePath, fileContent, { mode: 0o755 });

  //eslint-disable-next-line
  console.log(`‚úÖ DudeJS successfully installed ${hookName} hook for you.`);
};
