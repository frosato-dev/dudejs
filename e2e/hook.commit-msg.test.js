import path from "path";
import { exec, fileExists, read, write, cleanSrcDirectory, unifyTmpDir } from "./utils";
import { CLIENT_DIR } from "./constants";

const HOOK_PATH = path.join(CLIENT_DIR, ".git", "hooks", "commit-msg");

beforeAll(() => {
  cleanSrcDirectory();
});

it("should have a commit-msg file", () => {
  expect(fileExists(HOOK_PATH)).toBe(true);
  expect(read(HOOK_PATH)).toMatchSnapshot();
});

it("should pass with valid commit message", () => {
  write(`${CLIENT_DIR}/src/commit-lint.js`, `/*eslint-disable*/ console.log('hello commit-msg');`);

  exec("git add ./src/commit-lint.js");
  const stdout = exec('git commit -m "test: ci"');
  const value = stdout.toString();
  const toSnapshot = value.replace(/\[master \(root-commit\) [a-z0-9]{7}\]/, "<COMMIT_INFO>");
  expect(toSnapshot).toMatchSnapshot("commit-lint success");
});

it("should fail with wrong commit message", () => {
  write(`${CLIENT_DIR}/src/commit-lint.js`, `/*eslint-disable*/ console.log('hello commit-msg2');`);

  exec("git add ./src/commit-lint.js");
  let stdout;
  try {
    exec('git commit -m "foo: ci"');
  } catch (err) {
    stdout = err.stderr.toString();
  }

  expect(unifyTmpDir(stdout.replace(/[0-9]+ms/g, "XXms"))).toMatchSnapshot(
    "commit-lint error output",
  );
});
