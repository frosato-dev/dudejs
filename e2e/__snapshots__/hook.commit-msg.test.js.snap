// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should fail with wrong commit message: commit-lint error output 1`] = `
"src/commit-lint.js XXms
✅  Your files are fine dude!
⚠️️   Commit message : type must be one of [build, chore, ci, docs, feat, fix, perf, refactor, revert, style, test]
❌  Sorry dude, your commit message sucks
"
`;

exports[`should have a commit-msg file 1`] = `
"
#### DudeJS - START ####
npx dudejs msg \${1}
#### DudeJS - END ####
"
`;

exports[`should pass with valid commit message: commit-lint success 1`] = `
"<COMMIT_INFO> test: ci
 1 file changed, 1 insertion(+)
 create mode 100644 src/commit-lint.js
"
`;
